homeassistant/components/shelly/climate.py

```diff
@@ -49,21 +51,68 @@ async def async_setup_entry(
     if get_device_entry_gen(config_entry) == 2:
         return
 
+    wrapper: BlockDeviceWrapper = hass.data[DOMAIN][DATA_CONFIG_ENTRY][
+        config_entry.entry_id
+    ][BLOCK]
+
+    if wrapper.device.initialized:
+        await async_setup_climate_entities(async_add_entities, wrapper)
+    else:
+        await async_restore_climate_entities(
+            hass, config_entry, async_add_entities, wrapper
+        )
+
+
+async def async_setup_climate_entities(
+    async_add_entities: AddEntitiesCallback,
+    wrapper: BlockDeviceWrapper,
+) -> None:
+    """Set up online climate devices."""
+
+    _LOGGER.info("Setup online climate device %s", wrapper.name)
     device_block: Block | None = None
     sensor_block: Block | None = None
 
-    wrapper = hass.data[DOMAIN][DATA_CONFIG_ENTRY][config_entry.entry_id][BLOCK]
+    assert wrapper.device.blocks
+
     for block in wrapper.device.blocks:
         if block.type == "device":
             device_block = block
         if hasattr(block, "targetTemp"):
             sensor_block = block
 
     if sensor_block and device_block:
-        async_add_entities([ShellyClimate(wrapper, sensor_block, device_block)])
+        async_add_entities([BlockSleepingClimate(wrapper, sensor_block, device_block)])
 
 
-class ShellyClimate(ShellyBlockEntity, RestoreEntity, ClimateEntity):
+async def async_restore_climate_entities(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+    wrapper: BlockDeviceWrapper,
+) -> None:
+    """Restore sleeping climate devices."""
+    _LOGGER.info("Setup sleeping climate device %s", wrapper.name)
+
+    ent_reg = await entity_registry.async_get_registry(hass)
+    entries = entity_registry.async_entries_for_config_entry(
+        ent_reg, config_entry.entry_id
+    )
+
+    for entry in entries:
+
+        if entry.domain != CLIMATE_DOMAIN:
+            continue
+
+        _LOGGER.debug("Found entry %s [%s]", entry.original_name, entry.domain)
+        async_add_entities([BlockSleepingClimate(wrapper, None, None, entry)])
```

Comment:
Please collect all entities first and then make a single call to `async_add_entities`.